### PR TITLE
circuit for invalid creation code

### PR DESCRIPTION
**Pull request by DreamWuGit**
### Description

circuit for invalid creation code error in create, markdown spec https://github.com/privacy-scaling-explorations/zkevm-specs/pull/400
### Issue Link

issue #1291 

### Type of change

- [x] New feature (non-breaking change which adds functionality)

### Rationale

Get first byte of code to store , check it is 0xef

This PR contains:
- buss mapping: at the return opcode in create, get the first byte .
- add circuit & test 
- add tx deploy trace test .

### CODE REVIEW

**Comment by hero78119** [comment]
Hi @DreamWuGit , would you kindly resolved the conflict, since this pr https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1342 is merged :) ? Thanks!
I think few error handling step should be able to handled by new merged restore context function.

**Comment by lispc** [comment]
yes @DreamWuGit can copy new codes from scroll fork i think. i have already merged these branches.

**Comment by ed255** [comment]
@DreamWuGit kindly reminder: could you update this PR to resolve conflicts?  Thanks!

File: bus-mapping/src/evm/opcodes.rs [hunk 1]
```diff
@@ -253,7 +256,7 @@ fn fn_gen_associated_ops(opcode_id: &OpcodeId) -> FnGenAssociatedOps {
             DummySelfDestruct::gen_associated_ops
         }
         OpcodeId::CREATE => {
-            evm_unimplemented!("Using dummy gen_create_ops for opcode {:?}", opcode_id);
+            //evm_unimplemented!("Using dummy gen_create_ops for opcode {:?}", opcode_id);
```
[the remark this answers is not in the document]
Comment (DreamWuGit) [reply bus-mapping/src/evm/opcodes.rs:259]: Current status of this PR is that it depends on create gadget as following when it is merged, can remove the comment here.
- https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1356
- https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1357
- https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1358

**Comment by ChihChengLiang** [comment]
CREATE PRs (#1419 #1425 #1430 ) are all merged. This PR is unblocked

**Comment by lispc** [comment]
@silathdiir will work on this

**Comment by ChihChengLiang** [comment]
@silathdiir is this ready for review?

**Comment by silathdiir** [comment]
Yes, thanks. I update this PR with latest main branch, and also the spec doc PR https://github.com/privacy-scaling-explorations/zkevm-specs/pull/400.

Please help review when you have time. Thanks.

**Review by hero78119 — COMMENTED** [review]
Hi, one question regarding to soundness

In [create.rs](https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/main/zkevm-circuits/src/evm_circuit/execution/create.rs) and [return_revert](https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/main/zkevm-circuits/src/evm_circuit/execution/return_revert.rs), I didn't find the constraint where force the prover must fill `is_success == false` if code starting with `0xef`. I am wondering what if a malicious prover fill `is_success = true`, can the create/return_revert circuit catch up this InValidCreationWord error or actually it can pass?
From what I understand the only place to avoid this soundness is `create.rs` must also have constrain which force to go this execution step.

**Comment by silathdiir** [comment]
> Hi, one question regarding to soundness
> 
> In [create.rs](https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/main/zkevm-circuits/src/evm_circuit/execution/create.rs) and [return_revert](https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/main/zkevm-circuits/src/evm_circuit/execution/return_revert.rs), I didn't find the constraint where force the prover must fill `is_success == false` if code starting with `0xef`. I am wondering what if a malicious prover fill `is_success = true`, can the create/return_revert circuit catch up this InValidCreationWord error or actually it can pass? From what I understand the only place to avoid this soundness is `create.rs` must also have constrain which force to go this error execution step.

Yes, I see, I think you are right.
I will add `is_first_byte_invalid` and constrain as below to [create.rs](https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/main/zkevm-circuits/src/evm_circuit/execution/create.rs) and [begin_tx.rs](https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/main/zkevm-circuits/src/evm_circuit/execution/begin_tx.rs) (which may also call Create):
```
let is_first_byte_invalid = IsEqualGadget::construct(cb, first_byte.expr(), 0xef.expr());

cb.condition(is_success.expr(), |cb| {
    cb.require_zero(
        "init_code_first_byte != 0xef if success",
        is_first_byte_invalid
    );
});
```

**Comment by silathdiir** [comment]
Hi @hero78119, sorry for late update.

After some investigation, it seems that in CREATE (or CREATE2) execution step, [the contract Run result](https://github.com/ethereum/go-ethereum/blob/2754b197c935ee63101cbbca2752338246384fec/core/vm/evm.go#L465) could not be retrieved (from memory) to check [this ErrInvalidCode error](https://github.com/ethereum/go-ethereum/blob/2754b197c935ee63101cbbca2752338246384fec/core/vm/evm.go#L473). Please correct me if I was wrong, thanks.

So I fix to add a constraint which constrains the first byte of create init code must not be `0xef` to `return_revert` in this commit 3138114. Please help review again when you have time. Thanks.

File: zkevm-circuits/src/evm_circuit/util/common_gadget.rs [hunk 2]
```diff
@@ -970,13 +970,21 @@ impl<F: Field> CommonErrorGadget<F> {
         opcode: Expression<F>,
         rw_counter_delta: Expression<F>,
     ) -> Self {
-        cb.opcode_lookup(opcode.expr(), 1.expr());
+        Self::construct_with_last_return(cb, opcode, rw_counter_delta, 0.expr(), 0.expr())
+    }
 
-        let rw_counter_end_of_reversion = cb.query_cell();
+    pub(crate) fn construct_with_last_return(
```
Comment (hero78119) [inline zkevm-circuits/src/evm_circuit/util/common_gadget.rs:976]: how about function named as `construct_with_return_data`

File: bus-mapping/src/circuit_input_builder/input_state_ref.rs [hunk 3]
```diff
@@ -1064,7 +1064,8 @@ impl<'a> CircuitInputStateRef<'a> {
         let geth_step = steps
             .get(0)
             .ok_or(Error::InternalError("invalid index 0"))?;
-        let is_return_revert = geth_step.op == OpcodeId::REVERT || geth_step.op == OpcodeId::RETURN;
+        let is_return_revert = geth_step.op == OpcodeId::REVERT
+            || geth_step.op == OpcodeId::RETURN && exec_step.error.is_none();
 
         if !is_return_revert && !call.is_success {
```
[the remark this answers is not in the document]
Comment (silathdiir) [reply bus-mapping/src/circuit_input_builder/input_state_ref.rs]: Rename `is_return_revert` to `is_revert_or_return_call_success` in [66d6d42](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1292/commits/66d6d42bd8a9fe84444c7cd99b561e2175fab448#diff-1e6d053f6514ddc64497b6a584b3543a639730560b44b5fdf47a2c6b58b2ecd7L1067).

Will update the spec PR. Thanks.

File: zkevm-circuits/src/evm_circuit/execution/return_revert.rs [hunk 4]
```diff
@@ -318,9 +349,28 @@ impl<F: Field> ExecutionGadget<F> for ReturnRevertGadget<F> {
             .assign(region, offset, F::from(copy_rw_increase))?;
 
         let is_contract_deployment = call.is_create() && call.is_success && !length.is_zero();
+
+        let init_code_first_byte = if is_contract_deployment {
+            block.get_rws(step, 3).memory_value()
+        } else {
+            INVALID_INIT_CODE_FIRST_BYTE
```
[the remark this answers is not in the document]
Comment (silathdiir) [reply zkevm-circuits/src/evm_circuit/execution/return_revert.rs]: Set `init_code_first_byte` to 0 if else condition in commit [66d6d42](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1292/commits/66d6d42bd8a9fe84444c7cd99b561e2175fab448#diff-6fda08a70e867e93db76630eaa07d7bc725fb4b523a844c1803caac395695366L356).

File: zkevm-circuits/src/evm_circuit/util/common_gadget.rs (hunk 2 above)
Comment (silathdiir) [reply zkevm-circuits/src/evm_circuit/util/common_gadget.rs]: Rename `construct_with_last_return` to `construct_with_return_data` in commit [66d6d42](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1292/commits/66d6d42bd8a9fe84444c7cd99b561e2175fab448#diff-08d58ebc59ab6c11bf7c049e859441eed2319a03feb3fdffaddaba73f8a8c829L973).

**Review by hero78119 — APPROVED** [review]
👍 LGTM! 💯

**Comment by ChihChengLiang** [comment]
Hi @lispc,
Would you be available to review this PR?
We can assign another reviewer if you are busy.

**Comment by lispc** [comment]
i will have a look

**Review by lispc — APPROVED** [review]

**Pull request merged**